keras_hub/src/models/retinanet/retinanet_object_detector.py

```diff
@@ -0,0 +1,301 @@
+        head_dtype: Optional. The data type for the prediction heads.
+        prediction_decoder: Optional. A `keras.layers.Layer` that is
+            responsible for transforming RetinaNet predictions into usable
+            bounding box Tensors.
+            Defaults to `NonMaxSuppression` class instance.
+    """
+
+    backbone_cls = RetinaNetBackbone
+    preprocessor_cls = RetinaNetObjectDetectorPreprocessor
+
+    def __init__(
+        self,
+        backbone,
+        label_encoder,
+        anchor_generator,
+        num_classes,
+        bounding_box_format,
+        preprocessor=None,
+        activation=None,
+        head_dtype=None,
+        prediction_decoder=None,
+        **kwargs,
+    ):
+        # === Layers ===
+        head_dtype = head_dtype or backbone.dtype_policy
+        prior_probability = keras.initializers.Constant(
+            -1 * keras.ops.log((1 - 0.01) / 0.01)
+        )
+        box_head = PredictionHead(
+            output_filters=anchor_generator.anchors_per_location * 4,
+            num_conv_layers=4,
+            num_filters=256,
+            bias_initializer=prior_probability,
+            dtype=head_dtype,
+        )
+
+        classification_head = PredictionHead(
+            output_filters=anchor_generator.anchors_per_location * num_classes,
+            num_conv_layers=4,
+            num_filters=256,
+            dtype=head_dtype,
+        )
+
+        # === Functional Model ===
+        image_input = keras.layers.Input(backbone.image_shape, name="images")
+
+        feature_map = backbone(image_input)
+
+        cls_pred = []
+        box_pred = []
+        for level in feature_map:
+            box_pred.append(
+                keras.layers.Reshape((-1, 4))(box_head(feature_map[level]))
+            )
+            cls_pred.append(
+                keras.layers.Reshape((-1, num_classes))(
+                    classification_head(feature_map[level])
+                )
+            )
+
+        cls_pred = keras.layers.Concatenate(axis=1, name="classification")(
+            cls_pred
+        )
+        # box_pred is always in "center_yxhw" delta-encoded no matter what
+        # format you pass in.
+        box_pred = keras.layers.Concatenate(axis=1, name="box")(box_pred)
+
+        outputs = {"box": box_pred, "classification": cls_pred}
+
+        # === Config ===
+        self.bounding_box_format = bounding_box_format
+        self.num_classes = num_classes
+        self.backbone = backbone
+        self.preprocessor = preprocessor
+        self.label_encoder = label_encoder
+        self.anchor_generator = anchor_generator
+        self.activation = activation
+        self.box_head = box_head
+        self.classification_head = classification_head
+        self._prediction_decoder = prediction_decoder or NonMaxSuppression(
+            from_logits=(activation != keras.activations.sigmoid),
+            bounding_box_format=bounding_box_format,
+        )
+
+        super().__init__(
```

Comment:
move this code up to be with the functional model block

Reply:
Done